programs/cli_wallet/main.cpp

```diff
@@ -280,12 +280,18 @@ int main( int argc, char** argv )
          wallet_cli->register_api( wapi );
          wallet_cli->start();
 
-         fc::set_signal_handler([&wallet_cli, &wapi, &wallet_file](int signal) {
-            ilog( "Captured CTRL+C not in daemon mode" );
+         fc::set_signal_handler([&wapi, &wallet_file](int signal) {
+            ilog( "Captured SIGINT not in daemon mode" );
             wapi->save_wallet_file(wallet_file.generic_string());
-            wallet_cli->stop();
+            std::terminate();
```

Comment:
Terminate? Why not `exit`?

Reply:
because `exit` cannot finish program fully as `terminate` can

Reply:
I'm not convinced that we should call this. Actually, I don't like `exit` either, but IMHO `terminate` is even worse.

Reply:
```
# abort
indicates "abnormal" end to the program, and raises the the POSIX signal SIGABRT, which means that any handler that you have registered for that signal will be invoked, although the program will still terminate afterwords in either case. Usually you would use abort in a C program to exit from an unexpected error case where the error is likely to be a bug in the program, rather than something like bad input or a network failure. For example, you might abort if a data structure was found to have a NULL pointer in it when that should logically never happen.

# exit
indicates a "normal" end to the program, although this may still indicate a failure (but not a bug). In other words, you might exit with an error code if the user gave input that could not be parsed, or a file could not be read. An exit code of 0 indicates success. exit also optionally calls handlers before it ends the program. These are registered with the atexit and on_exit functions.

#std::terminate
is what is automatically called in a C++ program when there is an unhandled exception. This is essentially the C++ equivalent to abort, assuming that you are reporting all your exceptional errors by means of throwing exceptions. This calls a handler that is set by the std::set_terminate function, which by default simply calls abort.
```

Reply:
We should do a "normal" end.

Reply:
why we cannot use `terminate` here in programs/cli_wallet/main.cpp ? it isn't a library, we get SIGINT by capturing CTRL-C here and we need to fully terminate cli_wallet program, what is bad here ? before terminating we have to save wallet file correctly and after that we have to fully terminate cli_wallet program. What another idea/thoughts/solution do you see here instead of `terminate` call ?

Reply:
"normal" end won't fully terminate cli_wallet, we need to press CTRL-C twice (if using `exit` instead of `terminate`) in order to finish fully cli_wallet, is it right way ?

Reply:
Because both `terminate` and `abort` indicates an abnormal end of program. When this program is called by another program or is wrapped in a startup script, the external program or script may need to check the exit code of this program to perform subsequent jobs accordingly, in this case, an abnormal end will cause the external program / script to fail or act unexpectedly.

Reply:
ok understood about wrappers, didn't think about them yet, now and in future will know & think about them too

Reply:
> "normal" end won't fully terminate cli_wallet

This is what you need to figure out: how to end the program normally when needed.

Thanks.

Reply:
@cogutvalera I have explained above what the problem is, please try to solve instead of using dirty workarounds.

Reply:
@pmconrad I don't think `terminate` is dirty workaround, just didn't thought as @abitmore mentioned about other wrappers that may use cli_wallet program, you didn't told about wrappers and I didn't thought about them

Reply:
https://stackoverflow.com/questions/30250934/how-to-end-c-code

Reply:
Thanks !